src/EffectInfo.cpp

```diff
@@ -88,6 +93,52 @@ EffectBase* EffectInfo::CreateEffect(string effect_type) {
 	return NULL;
 }
 
+#if defined(__linux__)
+EffectBase* EffectInfo::LoadEffect(string location){
+    pthread_mutex_lock(&m_mutex);
+
+    void * file = dlopen(location.c_str(), RTLD_NOW);
+    if (file == nullptr){
+        pthread_mutex_unlock(&m_mutex);
+        return nullptr;
+    }
+    void * factory_ptr = dlsym(file, "factory");
+    if (factory_ptr == nullptr){
+        pthread_mutex_unlock(&m_mutex);
+        return nullptr;
```

Comment:
Shouldn't `file` be closed (er, `dlclose()`-d) before returning, here?

Reply:
I guess not, due to [documentation](https://linux.die.net/man/3/dlclose) if we call `dlclose`, this function (`factory`) will be unloaded.
In our case `dlclose` is being called only when you call `EffectInfo.UnloadDynamicEffects()`.

Reply:
> In our case `dlclose` is being called only when you call `EffectInfo.UnloadDynamicEffects()`.

Right, but that only happens if the reference returned by `dlopen()` is placed on the `m_loadedDynamicHandles` list — which **won't happen here**, because you return before reaching line 115.

That means you've got a dangling library handle hanging around, and the shared library won't be unloaded even when you _**do**_ later call `dlclose()`, because the handles are reference-counted and...
> a dynamic library is not deallocated until `dlclose()` has been called on it as many times as `dlopen()` has succeeded on it.

The code in the man page Example section doesn't `dlclose()` in its failure case because it's not calling `return()` — it's calling `exit()`. There's no point in cleaning up open references if you're exiting. As [Raymond Chen wrote](https://devblogs.microsoft.com/oldnewthing/?p=8683), on program exit "the building is being demolished. Don’t bother sweeping the floor and emptying the trash cans and erasing the whiteboards."

But if the code is expected to keep running past that point, then you don't want to be leaking `dlopen()` handles, it makes it impossible to clean up properly later on.

Reply:
@ferdnyc, yeah, I missed it, fully agree with you. I've added `dlclose()` in places where algorithm decides that this library is incorrect, but lib already has been loaded.